Redoes decal painter UI and adds tile sprayer

Tastyfish opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

### Changes
- Redoes the decal painter to give you a grid of previews to click instead of vague descriptions and dir combinations.
- Makes the UI come up on attack_self and changed dropping their cartridge on alt-click to match the behavior of most devices that let you do both.
  - Note the change to alt-click this applies to the airlock painter as well.

### Tile Sprayer
![image](https://user-images.githubusercontent.com/1185434/163708889-3b874b67-baff-41e5-bad2-a98fe95ee76c.png)
- Adds a tile sprayer, which has a slightly different icon.
  - The tile sprayer can apply the 4 corner colors, as well as the newer trimline fill edges.

### The new interfaces
![image](https://user-images.githubusercontent.com/1185434/163708792-fd3ec197-0ae4-43fc-9e99-defe19273131.png)


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

![image](https://user-images.githubusercontent.com/1185434/163709172-184407eb-47ab-4df3-9d5d-e9785a5634fa.png)

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Added a tile sprayer to paint the corner and edge colors of floors.
qol: Made the UI for decal painters more visual.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Tastyfish committed:
Redoes decal painter UI and adds tile sprayer

Fikou commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -229,21 +261,115 @@
+		list("Bar Burgundy", "#791500"),
+		list("Sec Red", "#DE3A3A"),
+		list("Cargo Brown", "#A46106"),
+		list("Engy Yellow", "#EFB341"),
```
```suggestion
		list("Engi Yellow", "#EFB341"),
```

Tastyfish replied:
Done.

Fikou commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -229,21 +261,115 @@
+		list("Sec Red", "#DE3A3A"),
+		list("Cargo Brown", "#A46106"),
+		list("Engy Yellow", "#EFB341"),
+		list("Botany Green", "#9FED58"),
```
```suggestion
		list("Service Green", "#9FED58"),
```

Tastyfish replied:
Done.

Tastyfish committed:
Suggested color palette name changes

Likteer commented:
https://user-images.githubusercontent.com/68878861/163710234-d97eb3d6-2af0-43f6-825c-3a34f3bc2dc2.mp4

stylemistake commented on tgui/packages/tgui/styles/interfaces/DecalPainter.scss:
```diff
@@ -0,0 +1,9 @@
+
+.DecalPainter__CellIcon {
+  display: block !important;
+  filter: drop-shadow(0px 0px 0px 2px black);
```
Why this and not box-shadow?

Tastyfish replied:
Ya know what? That dates back to me attempting a different style and I can just remove that as I found it doesn't even work in IE.

Tastyfish replied:
Removed.

stylemistake commented on tgui/packages/tgui/interfaces/DecalPainter.tsx:
```diff
@@ -0,0 +1,131 @@
+import { Window } from '../layouts';
+
+type Decal = {
+  name: string,
```
Types should be delimited by `;`, might be useful to format code with prettier (using vscode).

Tastyfish replied:
Fixed via prettier

stylemistake commented on tgui/packages/tgui/interfaces/DecalPainter.tsx:
```diff
@@ -0,0 +1,131 @@
+              return nondirectional ? (
+                // Tallll button for nondirectional
+                <IconButton
+                  decal={decal.decal} dir={2}
```
This is missing a key

Tastyfish replied:
Fixed

stylemistake commented on tgui/packages/tgui/interfaces/DecalPainter.tsx:
```diff
@@ -0,0 +1,131 @@
+              ) : (
+                // 4 buttons for directional
+                <Flex
+                  key="decal.decal"
```
I suppose this key is static because there's only one item in the list, but there should be some guarantees that this is a unique string in the list, because if there was more than one of this item with key "decal.decal", things would break horribly.

Tastyfish replied:
Fixed

stylemistake commented on tgui/packages/tgui/interfaces/DecalPainter.tsx:
```diff
@@ -0,0 +1,131 @@
+
+  return (
+    <Button
+      key={props.decal}
```
Key here does nothing. It should've been used directly on the element that's part of the list.

Tastyfish replied:
Oops, holdover from refactoring

Tastyfish replied:
Fixed

stylemistake reviewed (commented):
Some comments.

Onule commented:
Out of curiosity, would it be possible to have custom decal colors for the tiles?

Rocko1345 commented:
> Out of curiosity, would it be possible to have custom decal colors for the tiles?

I second this

Tastyfish commented:
> Out of curiosity, would it be possible to have custom decal colors for the tiles?

For the tile sprayer one, yes. I'll add this.

Tastyfish committed:
remove unused scss filter

Tastyfish committed:
prettier formatting of DecalPainter.tsc

Tastyfish committed:
added custom colors to tile sprayer

Tastyfish committed:
Prettier formatting

Tastyfish commented:
@stylemistake I fixed all the typescript issues mentioned (and one or 2 that I realized when reviewing it myself). As well, I added custom color support for the tile sprayer, necessitating some minimal refactoring for the extra button and special case icons.

stylemistake commented on tgui/packages/tgui/styles/interfaces/DecalPainter.scss:
```diff
@@ -0,0 +1,8 @@
+}
+
+.DecalPainter__CellIcon {
+  display: block !important;
```
It looks like you don't need this style since you can simply use a div for the icon, and that has display block by default.

stylemistake commented on tgui/packages/tgui/styles/interfaces/DecalPainter.scss:
```diff
@@ -0,0 +1,8 @@
+.DecalPainter__CellButton {
+  margin: 2px !important;
+  padding: 0.5em !important;
```
You can use `m={'2px'}` and `p={1}` properties on the Button to do the same thing. And not sure why did you use !important.

Tastyfish replied:
I'd think so, but for some reason they don't from a class perspective. I'll double check if m and p work properly in this case - I originally was trying to do this with table and it was a mess until I sensibly switched to a Flex

stylemistake replied:
Button implements a Box (and passes all unhandled props to it), so yes, it supports all of these props.

Tastyfish replied:
@stylemistake I didn't realize you could pass a string to `m` and `p` like that &mdash; neat. Problem with the DecalPainter__CellIcon one in the other syggestion: spritesheet css files make the icon elements `inline-block`, so the display has to be overriden. I've just made the icon div use `style={{display:'block'}}` since it's a single, simple property change.

stylemistake replied:
Ok, now I get why there was `display: block !important`. That kinda sucks.

stylemistake reviewed (approved):
LGTM, just a few more comments.

Tastyfish committed:
Removed needless classes

stylemistake commented on tgui/packages/tgui/interfaces/DecalPainter.tsx:
```diff
@@ -140,13 +140,14 @@ const IconButton = (props: IconButtonParams, context) => {
           dir: props.dir,
         })}>
-      <span className={`${icon} DecalPainter__CellIcon`} />
+      <div className={icon} style={{ display: 'block' }} />
```
Ok, now I get why there was `display: block` in the class, icon class was setting display to `inline-block`.

ArcaneMusic commented:
Gorgeous UI, thanks for taking it into the 21st century.

MrMelbert reviewed (changes requested):
Cool stuff

Tastyfish commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +148,251 @@
+	. = ..()
+	. += get_asset_datum(spritesheet_type)
+
+icon_assets
```
bizarre typo. no clue either but I guess I must have pasted it randomly at some late stage. whoops.

Tastyfish replied:
removed random line

Tastyfish committed:
Autodocs, minor fixes

Tastyfish commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -15,6 +15,7 @@
 	usesound = 'sound/effects/spray2.ogg'
 
 	var/obj/item/toner/ink = null
+	var/initial_ink_type = /obj/item/toner
```
done.

Tastyfish commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +148,251 @@
-			list("Delivery Marker","delivery"),
-			list("Warning Box","warn_full"))
+	var/spritesheet_type = /datum/asset/spritesheet/decals // spritesheet containing previews
+	var/supports_custom_color = FALSE // Add a custom color option
```
done. autodoc'd some existing vars from the old decal painter implementation as well

Tastyfish commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +148,251 @@
-	ui_interact(user)
+	var/turf/open/floor/floor_target = target
+	if(istype(floor_target, /turf/open/floor) && use_paint(user))
+		paint_floor(target)
```
Done. Forgot we had macros that specific for turfs, nice.

Tastyfish commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +148,251 @@
+	target.AddElement(/datum/element/decal, 'icons/turf/decals.dmi', stored_decal_total, stored_dir, null, null, alpha, color, null, CLEAN_TYPE_PAINT, null)
+
+/obj/item/airlock_painter/decal/proc/get_decal_path(decal, color, dir)
+	// Special case due to icon_state names
```
Done.

Tastyfish commented:
@MrMelbert I've done the suggestions and auto-doc'd all of the vars, and the new procs.

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	var/color_list = list(
+		list("Yellow", "yellow"),
+		list("Red", "red"),
+		list("White", "white"))
```
```suggestion
		list("White", "white"),
	)
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+		list("North", 1),
+		list("South", 2),
+		list("East", 4),
+		list("West", 8))
```
```suggestion
		list("West", 8),
	)
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+		list("White", "white"))
+	/// List of direction options as list(user-friendly label, dir value to return)
+	var/dir_list = list(
+		list("North", 1),
+		list("South", 2),
+		list("East", 4),
+		list("West", 8))
```
Shouldn't these be using `NORTH`, `SOUTH` etc, instead of 1/2/4/8?

Tastyfish replied:
Yeah, old code. changed.

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+		list("Box", "box"),
+		list("Box Corner", "box_corners"),
+		list("Delivery Marker", "delivery"),
+		list("Warning Box", "warn_full"))
```
```suggestion
		list("Warning Box", "warn_full"),
	)
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+		"bot",
+		"box",
+		"delivery",
+		"warn_full"
```
```suggestion
		"warn_full",
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	name = "floor_decals"
+	cross_round_cachable = TRUE
+
+	var/preview_floor_icon = 'icons/turf/floors.dmi'
+	var/preview_floor_state = "floor"
+	var/obj/item/airlock_painter/decal/painter_type = /obj/item/airlock_painter/decal
```
Worth documenting these too?

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+		list("Trimline Fill", "trimline_fill"),
+		list("Trimline Fill L", "trimline_fill__8"), // This is a hack that lives in the spritesheet builder and paint_floor
+		list("Trimline End", "trimline_end_fill"),
+		list("Trimline Box", "trimline_box_fill")
```
```suggestion
		list("Trimline Box", "trimline_box_fill"),
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	)
+	nondirectional_decals = list(
+		"tile_fourcorners",
+		"trimline_box_fill"
```
```suggestion
		"trimline_box_fill",
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	for(var/decal in decal_list)
+		.["decal_list"] += list(list(
+			"name" = decal[1],
+			"decal" = decal[2]
```
```suggestion
			"decal" = decal[2],
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	for(var/color in color_list)
+		.["color_list"] += list(list(
+			"name" = color[1],
+			"color" = color[2]
```
```suggestion
			"color" = color[2],
```

MrMelbert commented on code/game/objects/items/airlock_painter.dm:
```diff
@@ -146,104 +150,290 @@
+	for(var/dir in dir_list)
+		.["dir_list"] += list(list(
+			"name" = dir[1],
+			"dir" = dir[2]
```
```suggestion
			"dir" = dir[2],
```

MrMelbert reviewed (changes requested):
Last nitpicks. mostly indentation and trailing commas.

Tastyfish committed:
Did list format and spritesheet commenting suggestions

Tastyfish commented:
@MrMelbert In lieu of actually responding to every comment, I committed (and tested) all of the suggestions.
